pkg/cfg/flag.go

```diff
@@ -0,0 +1,115 @@
+package cfg
+
+import (
+	"flag"
+	"os"
+	"reflect"
+
+	"github.com/cortexproject/cortex/pkg/util/flagext"
+	ghodss "github.com/ghodss/yaml"
+	"github.com/pkg/errors"
+	yaml "gopkg.in/yaml.v2"
+)
+
+// FlagDefaults obtains defaults from a flag registerer and fails when it is not the first source.
+func FlagDefaults(reg flagext.Registerer, ptr *[]byte) Source {
+	f := FlagDefaultsDangerous(reg, ptr)
+
+	return func(dst interface{}) error {
+		// check dst is it's zero value (also indirect pointers first)
+		v := reflect.Indirect(reflect.ValueOf(dst))
+		z := reflect.Zero(v.Type())
+		if !reflect.DeepEqual(v.Interface(), z.Interface()) {
+			return errors.New("FlagDefaults() does not seem to be the first source. This is required to avoid data loss. If you really want to override data from previous sources, use FlagDefaultsDangerous()")
+		}
+
+		return f(dst)
+	}
+}
+
+// FlagDefaultsDangerous obtains defaults from a flag Registerer and DANGEROUSLY
+// replaces (!!) the dst object with reg.
+// Make sure this is always the first source, or it will overwrite other data.
+// The defaults are set into ptr, if is an empty byte slice and not nil.
+func FlagDefaultsDangerous(reg flagext.Registerer, ptr *[]byte) Source {
+	// some cortex code uses global `flag.*Var`. To avoid the flag package panicking, the default flagset is temporarily swapped out
+	tmp := flag.CommandLine
+	flag.CommandLine = flag.NewFlagSet("", flag.ContinueOnError)
+	flagext.DefaultValues(reg)
+	flag.CommandLine = tmp
+
+	// dump defaults for subsequent `Flags()` call
+	if ptr != nil {
+		d, err := yaml.Marshal(reg)
+		if err != nil {
+			panic(err)
+		}
+
+		*ptr = d
+	}
+
+	return func(dst interface{}) error {
+		// set *reg for *dst
+		v := reflect.Indirect(reflect.ValueOf(dst))
+		v.Set(reflect.Indirect(reflect.ValueOf(reg)))
+		return nil
+	}
+}
+
+// Flags returns a source that sets values from changed (!) flags onto dst.
+// The must have the same structure as dst, usually it is even the same type.
+// The defaults slice MUST contain valid defaults as yaml from the same registerer.
+func Flags(reg flagext.Registerer, defaults []byte) Source {
+	return dFlags(os.Args[1:], reg, defaults)
+}
+
+// dFlags parses flags defined by reg using the args slice onto dst.
+// The defaults slice MUST contain valid defaults as yaml from the same registerer.
+func dFlags(args []string, reg flagext.Registerer, defaults []byte) Source {
+	flagext.RegisterFlags(reg)
+	_ = flag.CommandLine.Parse(args)
+	data, err := yaml.Marshal(reg)
+	if err != nil {
+		panic(err)
+	}
+	return func(dst interface{}) error {
+		return stripDefaults(dst, data, defaults)
+	}
+}
+
+// stripDefaults takes the final config, the config from flags and the defaults,
+// but merges only non-default values onto dst
+func stripDefaults(dst interface{}, flagsStr, defStr []byte) error {
```

Comment:
What is happening here ? why do you need this ?

Reply:
The `flag` package gives us no way to distinguish between user-supplied and default-set flags. But to have correct precedence of the various configuration sources (supplied flags over yaml file over flag defaults), we need to be able to only apply the user supplied subset. 

This function takes both the supplied set (which also includes defaults) and only the defaults (which are computed by `FlagSet.Parse()` on an empty slice) in yaml marshalled form (for better passing around, because of pointer types, etc).

Everything that is not user-set (not different to the default) is removed then.

Reply:
However, while writing the explanation, I noticed that this in turn means, that it is not possible to user-set something to its default value. Example:

1. Flag `-verbose` defaults to false
2. Config file sets it to `true`
3. Specifying `-verbose=false` would be ignored

But I have no real idea how to address this issue

Reply:
Sad ! Not too sure what to think about this.

Reply:
Fixed it!